File: Python/OneLang/VM/ExprVM.py
from onelang_core import *
import OneLang.One.Ast.Expressions as exprs
import OneLang.VM.Values as vals

class VMContext:
    def __init__(self, model, hooks = None):
        self.model = model
        self.hooks = hooks

class ExprVM:
    def __init__(self, context):
        self.context = context
    
    def prop_access(self, obj, prop_name):
        if self.context.hooks != None:
            value = self.context.hooks.prop_access(obj, prop_name)
            if value != None:
                return value
        
        if not (isinstance(obj, vals.ObjectValue)):
            raise Error("You can only access a property of an object!")
        if not (prop_name in (obj).props):
            raise Error(f'''Property \'{prop_name}\' does not exists on this object!''')
        return (obj).props.get(prop_name)
    
    def evaluate(self, expr):
        if isinstance(expr, exprs.Identifier):
            return self.prop_access(self.context.model, expr.text)
        elif isinstance(expr, exprs.PropertyAccessExpression):
            obj_value = self.evaluate(expr.object)
            return self.prop_access(obj_value, expr.property_name)
        elif isinstance(expr, exprs.UnresolvedCallExpression):
            func = self.evaluate(expr.func)
            args = list(map(lambda x: self.evaluate(x), expr.args))
            result = func.call(args)
            return result
        elif isinstance(expr, exprs.StringLiteral):
            return vals.StringValue(expr.string_value)
        elif isinstance(expr, exprs.NumericLiteral):
            return vals.NumericValue(parse_int(expr.value_as_text))
        elif isinstance(expr, exprs.ConditionalExpression):
            cond_result = self.evaluate(expr.condition)
            result = self.evaluate(expr.when_true if (cond_result).value else expr.when_false)
            return result
        elif isinstance(expr, exprs.TemplateString):
            result = ""
            for part in expr.parts:
                if part.is_literal:
                    result += part.literal_text
                else:
                    value = self.evaluate(part.expression)
                    result += value.value if isinstance(value, vals.StringValue) else self.context.hooks.stringify_value(value)
            return vals.StringValue(result)
        elif isinstance(expr, exprs.BinaryExpression):
            left = self.evaluate(expr.left)
            right = self.evaluate(expr.right)
            if expr.operator == "==" or expr.operator == "===":
                return vals.BooleanValue(left.equals(right))
            elif expr.operator == "!=" or expr.operator == "!==":
                return vals.BooleanValue(not left.equals(right))
            else:
                raise Error(f'''Unsupported binary operator: {expr.operator}''')
        else:
            raise Error("Unsupported expression!")

File: PHP/src/OneLang/VM/ExprVM.php
<?php

namespace OneLang\VM\ExprVM;

use OneLang\One\Ast\Expressions\BinaryExpression;
use OneLang\One\Ast\Expressions\ConditionalExpression;
use OneLang\One\Ast\Expressions\Expression;
use OneLang\One\Ast\Expressions\Identifier;
use OneLang\One\Ast\Expressions\InstanceOfExpression;
use OneLang\One\Ast\Expressions\NumericLiteral;
use OneLang\One\Ast\Expressions\PropertyAccessExpression;
use OneLang\One\Ast\Expressions\StringLiteral;
use OneLang\One\Ast\Expressions\TemplateString;
use OneLang\One\Ast\Expressions\UnresolvedCallExpression;
use OneLang\VM\Values\BooleanValue;
use OneLang\VM\Values\ICallableValue;
use OneLang\VM\Values\IVMValue;
use OneLang\VM\Values\NumericValue;
use OneLang\VM\Values\ObjectValue;
use OneLang\VM\Values\StringValue;

interface IVMHooks {
    function propAccess($obj, $propName);
    
    function stringifyValue($value);
}

class VMContext {
    public $model;
    public $hooks;
    
    function __construct($model, $hooks = null) {
        $this->model = $model;
        $this->hooks = $hooks;
    }
}

class ExprVM {
    public $context;
    
    function __construct($context) {
        $this->context = $context;
    }
    
    function propAccess($obj, $propName) {
        if ($this->context->hooks !== null) {
            $value = $this->context->hooks->propAccess($obj, $propName);
            if ($value !== null)
                return $value;
        }
        
        if (!($obj instanceof ObjectValue))
            throw new \OneLang\Core\Error("You can only access a property of an object!");
        if (!(array_key_exists($propName, ($obj)->props)))
            throw new \OneLang\Core\Error("Property '" . $propName . "' does not exists on this object!");
        return @($obj)->props[$propName] ?? null;
    }
    
    function evaluate($expr) {
        if ($expr instanceof Identifier)
            return $this->propAccess($this->context->model, $expr->text);
        else if ($expr instanceof PropertyAccessExpression) {
            $objValue = $this->evaluate($expr->object);
            return $this->propAccess($objValue, $expr->propertyName);
        }
        else if ($expr instanceof UnresolvedCallExpression) {
            $func = $this->evaluate($expr->func);
            $args = array_map(function ($x) { return $this->evaluate($x); }, $expr->args);
            $result = $func->call($args);
            return $result;
        }
        else if ($expr instanceof StringLiteral)
            return new StringValue($expr->stringValue);
        else if ($expr instanceof NumericLiteral)
            return new NumericValue(Global.parseInt($expr->valueAsText));
        else if ($expr instanceof ConditionalExpression) {
            $condResult = $this->evaluate($expr->condition);
            $result = $this->evaluate(($condResult)->value ? $expr->whenTrue : $expr->whenFalse);
            return $result;
        }
        else if ($expr instanceof TemplateString) {
            $result = "";
            foreach ($expr->parts as $part) {
                if ($part->isLiteral)
                    $result .= $part->literalText;
                else {
                    $value = $this->evaluate($part->expression);
                    $result .= $value instanceof StringValue ? $value->value : $this->context->hooks->stringifyValue($value);
                }
            }
            return new StringValue($result);
        }
        else if ($expr instanceof BinaryExpression) {
            $left = $this->evaluate($expr->left);
            $right = $this->evaluate($expr->right);
            if ($expr->operator === "==" || $expr->operator === "===")
                return new BooleanValue($left->equals($right));
            else if ($expr->operator === "!=" || $expr->operator === "!==")
                return new BooleanValue(!$left->equals($right));
            else
                throw new \OneLang\Core\Error("Unsupported binary operator: " . $expr->operator);
        }
        else
            throw new \OneLang\Core\Error("Unsupported expression!");
    }
}
